File: bot.ts
import { Bot, Context, InlineKeyboard, Keyboard, session } from "grammy";
import { config } from "dotenv";
import { TEXTS, Lang } from "./i18n";
import { menu } from "./menu";
import { db, saveUser } from "./lib/db";
config();

export interface SessionData {
  step: "lang" | "fullname" | "contact" | "location" | "done";
  lang?: Lang;
  fullName?: string;
  phone?: string;
  address?: string;
  latitude?: number | null;
  longitude?: number | null;
  menuPath: string[];
}

type MyContext = Context & { session: SessionData };

const bot = new Bot<MyContext>(process.env.BOT_TOKEN!);

// Initialize session
bot.use(
  session({
    initial: (): SessionData => ({ step: "lang", menuPath: [] }),
  })
);

// Helper: Open main menu using the menu tree from menu.ts
async function openMenu(ctx: MyContext) {
  const lang = ctx.session.lang as Lang;

  ctx.session.menuPath = [];
  const mainMenuNode = menu[lang];
  const submenus = mainMenuNode.submenus || {};
  const keyboardOptions = Object.keys(submenus).map((key) => [key]);

  // Directly send the menu message with a persistent keyboard
  await ctx.reply(TEXTS[lang].main_menu, {
    reply_markup: {
      keyboard: keyboardOptions,
      resize_keyboard: true,
      one_time_keyboard: false, // Keep the keyboard persistent
    },
  });
}

// /start - if registration exists, skip to main menu; otherwise, start registration
bot.command("start", async (ctx) => {
  const userId = ctx.from?.id;
  if (userId) {
    const dbUser = await db.user.findUnique({
      where: { id: userId },
    });
    if (dbUser && dbUser.fullName && dbUser.phone) {
      // User already registered. Use session language if available, or default to "ru".
      ctx.session = {
        step: "done",
        lang: ctx.session.lang || "ru",
        fullName: dbUser.fullName,
        phone: dbUser.phone,
        address: dbUser.address || "",
        latitude: dbUser.latitude,
        longitude: dbUser.longitude,
        menuPath: [],
      };
      await openMenu(ctx);
      return;
    }
  }
  // Start registration flow.
  ctx.session.step = "lang";
  ctx.session.menuPath = [];
  const langKeyboard = new InlineKeyboard().text("UZ", "uz").text("RU", "ru");
  await ctx.reply("Select Language / Tilni tanlang / Выберите язык:", {
    reply_markup: langKeyboard,
  });
});

// Handle language selection (registration step: lang)
bot.callbackQuery(["uz", "ru"], async (ctx) => {
  if (ctx.session.step !== "lang") return;
  const lang = ctx.callbackQuery.data as Lang;
  ctx.session.lang = lang;
  ctx.session.step = "fullname";
  await ctx.answerCallbackQuery(
    `${TEXTS[lang].chosen_language} ${lang.toUpperCase()}`
  );
  await ctx.reply(TEXTS[lang].enter_full_name);
});

// Handle text messages for full name and menu navigation.
bot.on("message:text", async (ctx) => {
  const { step, lang } = ctx.session;
  if (!lang) return;

  // Registration: full name
  if (step === "fullname") {
    ctx.session.fullName = ctx.message.text;
    ctx.session.step = "contact";
    // Ask for phone using a reply keyboard that requests contact.
    const contactKeyboard = new Keyboard()
      .requestContact(TEXTS[lang].share_contact)
      .resized();
    await ctx.reply(TEXTS[lang].enter_phone, { reply_markup: contactKeyboard });
    return;
  }

  //handle location (adress) message for registration
  if (step === "location") {
    ctx.session.address = ctx.message.text;
    ctx.session.step = "done";
    // Save the registered user.
    if (ctx.from?.id) {
      await saveUser(ctx.from.id, ctx.session);
      await openMenu(ctx);
    }
    return;
  }

  // If already registered (step === done), treat text input as menu navigation.
  if (step === "done") {
    const text = ctx.message.text;

    // Start from the root of the menu tree.
    let currentNode = menu[lang];
    for (const key of ctx.session.menuPath) {
      if (!currentNode || !currentNode.submenus) {
        console.error("Invalid menu path:", ctx.session.menuPath);
        await ctx.reply(
          "Произошла ошибка в меню. Возвращаемся в главное меню."
        );
        await openMenu(ctx);
        return;
      }
      currentNode = currentNode.submenus[key];
    }

    // Handle "🛍 Заказать" -> Show categories
    if (text === "🛍 Заказать") {
      const categories = await db.category.findMany({
        where: { parentId: null }, // Fetch top-level categories
      });

      const categoryButtons = categories.map((category) => [category.name]);
      ctx.session.menuPath = ["🛍 Заказать"];
      await ctx.reply("Выберите категорию:", {
        reply_markup: {
          keyboard: [...categoryButtons, [TEXTS[lang].back]],
          resize_keyboard: true,
        },
      });
      return;
    }

    /// handle back button
    if (text === TEXTS[lang].back) {
      ctx.session.menuPath.pop();
      await openMenu(ctx);
      return;
    }

    // Handle category selection -> Show subcategories or products
    if (ctx.session.menuPath.includes("🛍 Заказать")) {
      const selectedCategory = await db.category.findUnique({
        where: { name: text },
        include: { subCategories: true, products: true },
      });

      if (!selectedCategory) {
        await ctx.reply("Категория не найдена.");
        return;
      }

      ctx.session.menuPath.push(text);

      if (selectedCategory.subCategories.length > 0) {
        // Show subcategories
        const subcategoryButtons = selectedCategory.subCategories.map((sub) => [
          sub.name,
        ]);
        await ctx.reply("Выберите подкатегорию:", {
          reply_markup: {
            keyboard: [...subcategoryButtons, [TEXTS[lang].back]],
            resize_keyboard: true,
          },
        });
      } else if (selectedCategory.products.length > 0) {
        // Show products
        const productButtons = selectedCategory.products.map((product) => [
          product.name,
        ]);
        await ctx.reply("Выберите продукт:", {
          reply_markup: {
            keyboard: [...productButtons, [TEXTS[lang].back]],
            resize_keyboard: true,
          },
        });
      } else {
        await ctx.reply("Нет доступных подкатегорий или продуктов.");
      }
      return;
    }

    // Handle product selection
    const selectedProduct = await db.product.findUnique({
      where: { name: text },
    });

    if (selectedProduct) {
      // Format the product details
      const productDetails = `
Продукт: ${selectedProduct.name}
Цена: ${selectedProduct.price}₽
${selectedProduct.description ? `Описание: ${selectedProduct.description}` : ""}
`;

      // Send the product details as a message
      await ctx.reply(productDetails, {
        reply_markup: {
          keyboard: [[TEXTS[lang].back]],
          resize_keyboard: true,
        },
      });
      return;
    }

    // Handle back button
    if (text === TEXTS[lang].back) {
      if (ctx.session.menuPath.length > 0) {
        ctx.session.menuPath.pop();
      }

      // Navigate back to the appropriate menu level
      currentNode = menu[lang];
      for (const key of ctx.session.menuPath) {
        currentNode = (currentNode.submenus || {})[key];
      }

      if (ctx.session.menuPath.includes("🛍 Заказать")) {
        const parentCategoryName =
          ctx.session.menuPath[ctx.session.menuPath.length - 1];
        const parentCategory = await db.category.findUnique({
          where: { name: parentCategoryName },
          include: { subCategories: true, products: true },
        });

        if (
          parentCategory?.subCategories &&
          parentCategory?.subCategories.length > 0
        ) {
          const subcategoryButtons = parentCategory.subCategories.map((sub) => [
            sub.name,
          ]);
          await ctx.reply("Выберите подкатегорию:", {
            reply_markup: {
              keyboard: [...subcategoryButtons, [TEXTS[lang].back]],
              resize_keyboard: true,
            },
          });
        } else if (
          parentCategory?.subCategories &&
          parentCategory?.subCategories.length > 0
        ) {
          const productButtons = parentCategory.products.map((product) => [
            product.name,
          ]);
          await ctx.reply("Выберите продукт:", {
            reply_markup: {
              keyboard: [...productButtons, [TEXTS[lang].back]],
              resize_keyboard: true,
            },
          });
        } else {
          await ctx.reply("Выберите категорию:", {
            reply_markup: {
              keyboard: [[TEXTS[lang].back]],
              resize_keyboard: true,
            },
          });
        }
      } else {
        // Return to the main menu
        await openMenu(ctx);
      }
      return;
    }

    // If input doesn't match any menu item, notify the user
    await ctx.reply("Пожалуйста, выберите пункт меню.");
  }
});

// Handle contact message for registration (phone)
bot.on("message:contact", async (ctx) => {
  const { step, lang } = ctx.session;
  if (!lang) return;
  if (step === "contact") {
    ctx.session.phone = ctx.message.contact.phone_number;
    ctx.session.step = "location";
    // Ask for location using a reply keyboard that requests location.
    const locationKeyboard = new Keyboard()
      .requestLocation("Share Location")
      .resized();
    await ctx.reply("Please share your location:", {
      reply_markup: locationKeyboard,
    });
    return;
  }
});

// Handle location message for registration.
bot.on("message:location", async (ctx) => {
  const { step, lang } = ctx.session;
  if (!lang) return;
  if (step === "location") {
    ctx.session.latitude = ctx.message.location.latitude;
    ctx.session.longitude = ctx.message.location.longitude;
    ctx.session.step = "done";
    // Save the registered user.
    if (ctx.from?.id) {
      await saveUser(ctx.from.id, ctx.session);
    }
    await openMenu(ctx);
    return;
  }
});

bot.api.setMyCommands([
  {
    command: "start",
    description: "Запуск бота | Botni ishga tushirish",
  },
  {
    command: "help",
    description: "Помощь | Yordam",
  },
]);

bot.catch((err) => {
  console.error("Bot error:", err);
});

bot.start();

File: bot.py
import logging
from telegram import ForceReply, KeyboardButton, ReplyKeyboardMarkup, Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, ContextTypes, MessageHandler, filters, CallbackQueryHandler

# Enable logging
logging.basicConfig(
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s", level=logging.INFO
)
logging.getLogger("httpx").setLevel(logging.WARNING)

logger = logging.getLogger(__name__)

TEXTS = {
    'ru': {
        'chosen_language': "Выберите язык:",
        'enter_full_name': "Введите ваше имя и фамилию:",
        'enter_phone': "Введите ваш номер телефона:",
        'main_menu': "Главное меню:",
        'main_buttons': [["🛍 Заказать", "🔥 Акция"], ["📦 Мои заказы", "⚙️ Настройки"], [ "📍 Ближайший филиал"]],
        'back': "⬅️ Назад",
        'order_menu': ['Доставка', 'Самовывоз'],
        'share_contact': "📞 Отправить номер телефона",
        'help': "Телефона",
    },
    'uz': {
        'chosen_language': "Tanlangan til:",
        'enter_full_name': "Ismingiz va familiyangizni kiriting:",
        'enter_phone': "Telefon raqamingizni kiriting:",
        'main_menu': "Asosiy menyu:",
        'main_buttons': [["🛍 Buyurtma berish", "🔥 Aksiya"], [ "📦 Buyurtmalarim", "⚙️ Sozlamalar"], [ "📍 Eng yaqin filial"]],
        'back': "⬅️ Orqaga",
        'order_menu': ['Yetkazib berish', 'Olib ketish'],
        'share_contact': "📞 Telefon raqamni yuborish",
        'help': "Yordam",
    }
}

user = {}  # Default language

# Define a few command handlers
async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Send a message when the command /help is issued."""
    user_id = update.effective_user.id
    lang = user.get(user_id, {}).get('language', 'ru')  # Default to 'ru' if no language selected
    await update.message.reply_text(TEXTS[lang]['help'])


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Send a message when the command /start is issued."""
    
    user_id = update.effective_user.id
    user[user_id] = {'language': None, 'full_name': None, 'phone': None}
    
    keyboard = [
        [
            InlineKeyboardButton("🇷🇺 Русский", callback_data='ru'),
            InlineKeyboardButton("🇺🇿 O'zbek", callback_data='uz'),
        ]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)

    await update.message.reply_text(
        TEXTS['ru']['chosen_language'] + "\n" + TEXTS['uz']['chosen_language'],
        reply_markup=reply_markup,
    )


# Handle language selection callback query
async def button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle the button press for language selection."""
    query = update.callback_query
    user_id = query.from_user.id
    selected_language = query.data
    
    # Save the selected language in the user dictionary
    user[user_id]['language'] = selected_language
    
    # Send a confirmation message based on the selected language
    lang_texts = TEXTS[selected_language]
    await query.answer()  # Acknowledge the callback
    # Update the keyboard to show a checkmark next to the selected language
    keyboard = [
        [
            InlineKeyboardButton(
                "🇷🇺 Русский ✅" if selected_language == 'ru' else "🇷🇺 Русский", 
                callback_data='ru'
            ),
            InlineKeyboardButton(
                "🇺🇿 O'zbek ✅" if selected_language == 'uz' else "🇺🇿 O'zbek", 
                callback_data='uz'
            ),
        ]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    # Edit the message to update the keyboard
    await query.edit_message_text(
        lang_texts['chosen_language'], reply_markup=reply_markup
    )

    # You can then continue with additional steps, like asking for name or phone number
    user[user_id]['awaiting'] = 'full_name'  # Set state to await full name
    await query.message.reply_text(lang_texts['enter_full_name'])

# New handler to capture full name and phone number messages
async def handle_text(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Handle messages for full name and phone input."""
    user_id = update.effective_user.id
    if user_id not in user:
        await update.message.reply_text("Please send /start to begin the registration process.")
        return

    user_state = user[user_id].get('awaiting')
    lang = user[user_id].get('language', 'ru')
    
    if user_state == 'full_name':
        # Save the full name and update state to wait for phone number
        user[user_id]['full_name'] = update.message.text
        user[user_id]['awaiting'] = 'phone'
        await update.message.reply_text(
            TEXTS[lang]['enter_phone'],
            reply_markup=ReplyKeyboardMarkup(
            [[KeyboardButton(TEXTS[lang]['share_contact'], request_contact=True)]],
            resize_keyboard=True,
            one_time_keyboard=True
            )
        )
    elif user_state == 'phone':
        if update.message.contact:
            # Save the phone number from the shared contact
            user[user_id]['phone'] = update.message.contact.phone_number
        else:
            # Save the phone number from the text input
            user[user_id]['phone'] = update.message.text
        user[user_id].pop('awaiting', None)
        # Registration complete, show the main menu
        main_menu_buttons = TEXTS[lang]['main_buttons']
        await update.message.reply_text(
            TEXTS[lang]['main_menu'],
            reply_markup=ReplyKeyboardMarkup(
            main_menu_buttons,
            resize_keyboard=True
            )
        )

    else:
        await update.message.reply_text("I'm not expecting any input right now.")
    
# Handle button presses for "🛍 Buyurtma berish" or "🛍 Заказать"
async def handle_order_button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user_id = update.effective_user.id
    lang = user[user_id].get('language', 'ru')
    order_buttons = [[TEXTS[lang]['back']]] + [[order_type] for order_type in TEXTS[lang]['order_types']]
    await update.message.reply_text(
        TEXTS[lang]['main_menu'],
        reply_markup=ReplyKeyboardMarkup(
            order_buttons,
            resize_keyboard=True
        )
    )

# Handle button presses for "🛍 Buyurtma berish" or "🛍 Заказать"
async def handle_order_button(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user_id = update.effective_user.id
    lang = user[user_id].get('language', 'ru')
    order_buttons = [[TEXTS[lang]['back']]] + [[order_type] for order_type in TEXTS[lang]['order_types']]
    await update.message.reply_text(
    TEXTS[lang]['main_menu'],
    reply_markup=ReplyKeyboardMarkup(
        order_buttons,
        resize_keyboard=True
    )
    )

    
def main() -> None:
    """Start the bot."""
    # Create the Application and pass it your bot's token.
    application = Application.builder().token("7820685548:AAFllspW0dGp6xDljt3NpoC_iI7mdSgHGA8").build()

    # on different commands - answer in Telegram
    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("help", help_command))

    # Handle callback queries (button presses)
    application.add_handler(CallbackQueryHandler(button))

    # Handle text messages for full name and phone number
    application.add_handler(MessageHandler(filters.CONTACT, handle_text))
    # Handle reply keyboard button presses
    application.add_handler(MessageHandler(filters.Regex("⬅️ Назад|⬅️ Orqaga"), handle_text))
    
    # Add a handler for the order button
    application.add_handler(MessageHandler(filters.Regex("🛍 Buyurtma berish|🛍 Заказать"), handle_order_button))

    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_text))

    # Run the bot until the user presses Ctrl-C
    application.run_polling(allowed_updates=Update.ALL_TYPES)

if __name__ == "__main__":
    main()
